### DISCUSSION TITLE
Proper way to do thermal radiation with raytracing

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#18874] [Q&A General] [answered]
Hi All

I think im doing this right, I have an unstructured mesh, with three sidesets listed with the elucidating names thermal-rad-1 through thermal-rad-3. I have a GrayLambertNeumannBC specification
```
  [radiation]
    type = GrayLambertNeumannBC
    variable = temp
    reconstruct_emission = false
    surface_radiation_object_name = gray_lambert
    boundary = 'thermal-rad-1 thermal-rad-2 thermal-rad-3'
  []
```
I got an error from Using the GrayDiffuseRadiation - https://mooseframework.inl.gov/syntax/GrayDiffuseRadiation/ example radiative_transfer_action.i when using an unstructured mesh, telling me that it only works with GeneratedMeshes rather than FileMeshes. So instead, I created the following
```
[UserObjects]
  [gray_lambert]
    type = ViewFactorObjectSurfaceRadiation
    boundary = 'thermal-rad-1 thermal-rad-2 thermal-rad-3'
    adiabatic_boundary = 'thermal-rad-1 thermal-rad-2 thermal-rad-3'
    emissivity = '0.3 0.9 0.1'
    temperature = temp
    view_factor_object_name = view_factor
  []
  [view_factor_study]
    type = ViewFactorRayStudy
    execute_on = initial
    boundary = 'thermal-rad-1 thermal-rad-2 thermal-rad-3'
    face_order = FOURTH
    polar_quad_order = 12
    azimuthal_quad_order = 4
    ray_tracing_face_type = GAUSS
    warn_subdomain_hmax = false
    view_factor_calculator = ray_tracing
  []
  [view_factor]
    type = RayTracingViewFactor
    boundary = 'thermal-rad-1 thermal-rad-2 thermal-rad-3'
    execute_on = INITIAL
    ray_study_name = view_factor_study
  []
[]

[RayBCs/viewfactor]
  type = ViewFactorRayBC
  boundary = 'thermal-rad-1 thermal-rad-2 thermal-rad-3'
[]
```
However, I get an error 
```

*** ERROR ***
The following error occurred in the object "view_factor_study", of type "ViewFactorRayStudy".

Cannot use GRID quadrature type with tetrahedral elements in ViewFactorRayStudy 'view_factor_study'
```
You can see above that ray_tracing_face_type is set to Gauss, so I'm at a loss, is this a genuine bug?

**Comment by GiudGiud** [comment]
@loganharbour @snschune

**Comment by snschune** [comment]
A few general things, I would use the action to set up this problem. Look for `[GrayDiffuseRadiation]` in the heat conduction module's input. Unstructured meshes are supported but triangles do not work for the grid quadrature. There should be a quadrature type that works with triangles though (I am saying triangles because that's what the sidesets around tets are made of).

I am not sure if it makes sense to have all surfaces be adiabatic:
```
boundary = 'thermal-rad-1 thermal-rad-2 thermal-rad-3'
adiabatic_boundary = 'thermal-rad-1 thermal-rad-2 thermal-rad-3'
```
Do you have a picture of the geometry by any chance? 

The general way net radiation transport works in MOOSE is on a cavity basis. A cavity is a meshed region of your domain that is fully enclosed by sidesets that are listed under the `boundary` parameter in the `[GrayDiffuseRadiation]` block. There are three types of sidesets:
1. `fixed_temperature_boundary` keep the temperature constant, no heat conduction variable is defined "behind" that sideset.
2. `adiabatic_boundary` radiation in equals radiation out, no heat conduction variable is defined "behind" that sideset.
3. Anything in `boundary` that isn't 1 or 2 has a block "behind" it that defines the `MooseVariable` that you provide in the `temperature` keyword. 
You can define more than one cavity.

**Comment by makeclean** [comment]
Ok, I've got some progress 
```
[GrayDiffuseRadiation]
  [./cavity]
    boundary = 'thermal-rad-1 thermal-rad-2 thermal-rad-3'
    emissivity = '0.7 0.1 0.3'
    n_patches = '40 20 30'
    partitioners = 'metis metis metis'
    temperature = temp
    view_factor_calculator = ray_tracing
    ray_tracing_face_type = GAUSS
  [../]
[]
```
I get the same error as I did on the last message on this in may. I now get the error
```
*** ERROR ***
Ray on processor 0 and thread 0 failed to trace at line 2145

Don't know what to do with a Ray after it hit an external
boundary at point (x,y,z)=(  0.0178, 0.00248967, -0.0147817)!

When hitting an external RayBC, a Ray must either:
  Be killed by a RayBC
  Have its trajectory changed by the RayBC
by at least one of the executed RayBCs.

You need to either:
  Kill/change the Ray sooner with RayKernels, internal RayBCs, or a max distance
  Kill/change the Ray on the boundary with a RayBC
```
The geometry looks like this;
![Screenshot from 2021-09-21 13-47-07](https://user-images.githubusercontent.com/2439754/134173097-b42f48c8-efb8-426b-bc1d-24fc6083b8c3.png)
The yellow region in the middle is my cavity (the mesh is sliced so you can see it (in reality its whole), there are no gaps in the sidesets, all the triangles in the 3 sets define a closed boundary. So I'm at a loss here.

**Reply by makeclean** [reply]
I also thought that the block which defines the mesh of the vacuum region, would automatically not need a kernel to be active?

**Reply by snschune** [reply]
That error typically happens when the sidesets do not completely enclose the cavity. I would usually load the geometry into paraview (you can create exodus versions of your mesh using MOOSE using --mesh-only command line option), make all blocks invisible and only show the three sidesets. If that doesn't help, I'd use a sphere located at `0.0178, 0.00248967, -0.0147817` in paraview to see where the ray fails to trace.

**Reply by makeclean** [reply]
Ok, I've reproduced a more simple problem, and still no-dice. I now have a cube inside of a cube, of side 10 m, and the smaller cube 2m. I have, what I consider to be a minimum (non) working example. The CAD does not overlap, the geometry meshes as it should. I have 6 side sets involved in the radiative transfer, named cav-1 through cav-6. I'm happy to attach meshes and cubit files if that helps. 
```
[Mesh]
  [mesh-in]
    type = FileMeshGenerator
    file = 'heat-ray.e'
  []
[]

[Variables]
  [temp]
    initial_condition = 313.15
  []
[]


[AuxVariables]
  [temp_in_C]
  []
  [radiation_flux]
  []
[]

[Kernels]
  [isotropic-heat]
    type = HeatConduction
    variable = temp
    block = 'brik'
  []
  [null-kernel]
    type = NullKernel
    variable = temp 
    block = 'vacuum'
  []
[]

[AuxKernels]
  [K_to_C]
    type = ParsedAux
    variable = 'temp_in_C'
    function = 'temp-273.15'
    args = 'temp'
  []
[]

[GrayDiffuseRadiation]
  [./cavity]
    boundary = 'cav-1 cav-2 cav-3 cav-4 cav-5 cav-6'
    emissivity = '0.5 0.5 0.5 0.5 0.5 0.5'
    n_patches = '20 20 20 20 20 20'
    partitioners = 'metis metis metis metis metis metis'
    temperature = temp
    view_factor_calculator = ray_tracing
    ray_tracing_face_type = GAUSS
  [../]
[]

[BCs]
  [fixed-temp-1]
    type = DirichletBC
    variable = temp
    boundary = 't1'
    value = 313.15
  []
  [fixed-temp-2]
    type = DirichletBC
    variable = temp
    boundary = 't2'
    value = 900.0
  []
[]
        
[Materials]
  [heat-sink_density]
    type = GenericConstantMaterial
    prop_names = 'density'
    block = 'brik'
    prop_values = '1.1e3'
  []
  [heat-sink]
    type = HeatConductionMaterial
    block = 'brik'
    specific_heat = 200
    thermal_conductivity = 50
  []
  [vacuum_density]
    type = GenericConstantMaterial
    prop_names = 'density'
    block = 'vacuum'
    prop_values = '0'
  []
  [vacuum]
    type = HeatConductionMaterial
    block = 'vacuum'
    specific_heat = 0
    thermal_conductivity = 0
  []
[]

[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  automatic_scaling = true
  solve_type = 'NEWTON'
  type = Steady
  line_search = none
  nl_abs_tol = 1e-6
  nl_rel_tol = 1e-5
  l_tol = 1e-4

  l_max_its = 100
  nl_max_its = 10
  dt = 0.1
  num_steps = 30

  petsc_options_iname = '-pc_type -ksp_gmres_restart'
  petsc_options_value = 'ilu      101'

[]
  
[Outputs]
  exodus = true
[]
```
The error is, 
```
*** ERROR ***
Ray on processor 0 and thread 0 failed to trace at line 2145

Don't know what to do with a Ray after it hit an external
boundary at point (x,y,z)=(       5, -1.02841, -0.694079)!

When hitting an external RayBC, a Ray must either:
  Be killed by a RayBC
  Have its trajectory changed by the RayBC
by at least one of the executed RayBCs.

You need to either:
  Kill/change the Ray sooner with RayKernels, internal RayBCs, or a max distance
  Kill/change the Ray on the boundary with a RayBC
```
Now my understanding is I shouldnt have to define more BC's, Im happy to, but the documentation indicates that this problem *should* work - what am I doing wrong?

**Reply by makeclean** [reply]
Here are the files - https://www.dropbox.com/sh/2f0odxinfcxivkj/AAAM0ZYfULUCUeunoCC5WZdla?dl=0

**Reply by snschune** [reply]
Here is a slight modification of the input file that runs. The main difference is how the sidesets are defined. It's a bit complicated so bear with me. 

Sidesets are a collection of faces, where a face is what is around elements (triangle that bound a tet, or quadrilaterals that bound a hex). Faces do not exist on their own in MOOSE, they are indexed by two parameters `element id, local side id`. So for a hexahedron, you would have 8 local side ids from 0 to 7.

Now, you'd observe immediately, that interior faces have two ways of addressing them, either from the left or the right element bordering them. These two ways of defining a face are actually not interchangeable in MOOSE. The difference is which direction the normal points. Let's say element 1 borders element 2 and the common face has side id 0 for element 1 and side id 3 for element 2. Then the face can be indexed by `elem 1, side 0` or `elem 2, side 3`. The difference between these two alternatives is that for case 1, the normal points from element 1 to 2, while the reverse is true for the second definition. For the first definition, the face is grouped with elem 1, while for the second definition the side is grouped with elem 2. That carries through if you e.g. delete blocks and elem 1 is not deleted, but elem 2 is deleted etc.

What I've essentially done is to make sure that the faces making up the sideset `inbetween` is defined on the `brik` side. 
The rule is that sidesets that participate in the net radiation transfer and have temperature defined in the block "behind"
them, must be defined using the element/local side definition that belongs to the block that you solve temperature on.
This ensures that the normal points into the cavity.

For isothermal and adiabatic sidesets, the idea usually is that these are external surfaces anyway so their normals are usually pointing out of the cavity (nothing you can do because there is no neighboring element at an external boundary).  

So what you have to do is to flip normals in CUBIT for all the internal sidesets that participate in the net radiation transport calculation. 


```
[Mesh]
  [mesh-in]
    type = FileMeshGenerator
    file = 'heat-ray.e'
  []

  [sidesets]
    type = SideSetsBetweenSubdomainsGenerator
    primary_block = 2
    paired_block = 1
    new_boundary = 101
    input = mesh-in
  []

  [name_sideset]
    type = RenameBoundaryGenerator
    old_boundary = 101
    new_boundary = inbetween
    input = sidesets
  []
[]

[Variables]
  [temp]
    initial_condition = 313.15
  []
[]


[AuxVariables]
  [temp_in_C]
  []
  [radiation_flux]
  []
[]

[Kernels]
  [isotropic-heat]
    type = HeatConduction
    variable = temp
    block = 'brik'
  []
  [null-kernel]
    type = NullKernel
    variable = temp
    block = 'vacuum'
  []
[]

[AuxKernels]
  [K_to_C]
    type = ParsedAux
    variable = 'temp_in_C'
    function = 'temp-273.15'
    args = 'temp'
  []
[]

[GrayDiffuseRadiation]
  [./cavity]
    #boundary = 'cav-1 cav-2 cav-3 cav-4 cav-5 cav-6'
    #emissivity = '0.5 0.5 0.5 0.5 0.5 0.5'
    #n_patches = '20 20 20 20 20 20'
    #partitioners = 'metis metis metis metis metis metis'
    boundary = 'inbetween'
    emissivity = '0.5'
    n_patches = '100'
    partitioners = 'metis'
    temperature = temp
    view_factor_calculator = ray_tracing
    ray_tracing_face_type = GAUSS
  [../]
[]

[BCs]
  [fixed-temp-1]
    type = DirichletBC
    variable = temp
    boundary = 't1'
    value = 313.15
  []
  [fixed-temp-2]
    type = DirichletBC
    variable = temp
    boundary = 't2'
    value = 900.0
  []
[]

[Materials]
  [heat-sink_density]
    type = GenericConstantMaterial
    prop_names = 'density'
    block = 'brik'
    prop_values = '1.1e3'
  []
  [heat-sink]
    type = HeatConductionMaterial
    block = 'brik'
    specific_heat = 200
    thermal_conductivity = 50
  []
  [vacuum_density]
    type = GenericConstantMaterial
    prop_names = 'density'
    block = 'vacuum'
    prop_values = '0'
  []
  [vacuum]
    type = HeatConductionMaterial
    block = 'vacuum'
    specific_heat = 0
    thermal_conductivity = 0
  []
[]

[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  automatic_scaling = true
  solve_type = 'NEWTON'
  type = Steady
  line_search = none
  nl_abs_tol = 1e-6
  nl_rel_tol = 1e-5
  l_tol = 1e-4

  l_max_its = 100
  nl_max_its = 10

  petsc_options_iname = '-pc_type -ksp_gmres_restart'
  petsc_options_value = 'ilu      101'

[]

[Outputs]
  exodus = true
[]
```

**Comment by snschune** [comment, marked as answer]
Small addition here, you see that I name boundary 101. That is necessary due to this bug (https://github.com/idaholab/moose/issues/17192) that will be fixed once we switched
to id based identification of patches. 

```
  [sidesets]
    type = SideSetsBetweenSubdomainsGenerator
    primary_block = 2
    paired_block = 1
    new_boundary = 101
    input = mesh-in
  []

  [name_sideset]
    type = RenameBoundaryGenerator
    old_boundary = 101
    new_boundary = inbetween
    input = sidesets
  []
```

**Reply by makeclean** [reply]
No, that makes complete sense - I was trying to achieve the same thing Cubit, which was proving much more difficult :)

**Reply by makeclean** [reply]
However, the NullKernel thing still remains, should that still be happening, do I need to add a null kernel to my vacuum region?

**Reply by GiudGiud** [reply]
you can use kernel_coverage_check = false in [Problem] to not require a NullKernel
A NullKernel should in general be avoided, that will create a bunch of 0s in your system matrix for the relevant dofs iirc
